swift/log/SwiftDiagnostics.cpp

```diff
@@ -0,0 +1,73 @@
+#include "swift/log/SwiftDiagnostics.h"
+
+#include <date/date.h>
+#include <binlog/Entries.hpp>
+#include <nlohmann/json.hpp>
+
+namespace codeql {
+SwiftDiagnosticsSource::SwiftDiagnosticsSource(std::string_view internalId,
+                                               std::string&& name,
+                                               std::vector<std::string>&& helpLinks,
+                                               std::string&& action)
+    : name{std::move(name)}, helpLinks{std::move(helpLinks)}, action{std::move(action)} {
+  id = extractorName;
+  id += '/';
+  id += programName;
+  id += '/';
+  std::transform(internalId.begin(), internalId.end(), std::back_inserter(id),
+                 [](char c) { return c == '_' ? '-' : c; });
+}
+
+void SwiftDiagnosticsSource::create(std::string_view id,
+                                    std::string name,
+                                    std::vector<std::string> helpLinks,
+                                    std::string action) {
+  auto [it, inserted] = map().emplace(
+      id, SwiftDiagnosticsSource{id, std::move(name), std::move(helpLinks), std::move(action)});
+  assert(inserted);
+}
+
+void SwiftDiagnosticsSource::emit(std::ostream& out,
+                                  std::string_view timestamp,
+                                  std::string_view message) const {
+  nlohmann::json entry;
```

Comment:
Optional: It's [possible to initialise](https://github.com/nlohmann/json#json-as-first-class-data-type) a `nlohmann::json` object directly, like:

```
nlohmann::json entry = {
  {"source", {
    {"id", id},
    {"name", name},
    {"extractorName", extractorName}
  },
  ...
```

I think that's slightly more readable particularly in the context of emitting JSON (since it looks more like JSON).